File: src/Modules/Prescriptions/ErpClink.Modules.Prescriptions.Application/Prescriptions/Models/PrescriptionModels.cs
namespace ErpClink.Modules.Prescriptions.Application.Prescriptions.Models;

public sealed record PrescriptionItemInput(
    Guid MedicationId,
    string Dosage,
    string Frequency,
    string? Duration,
    string? Route,
    string? Instructions,
    decimal? Quantity,
    string? Notes,
    int? SortOrder);

public sealed record CreatePrescriptionRequest(
    Guid MedicalVisitId,
    string? Notes,
    IReadOnlyList<PrescriptionItemInput>? Items);

public sealed record UpdatePrescriptionRequest(string? Notes, byte[]? RowVersion);

public sealed record UpdatePrescriptionItemRequest(
    string Dosage,
    string Frequency,
    string? Duration,
    string? Route,
    string? Instructions,
    decimal? Quantity,
    string? Notes,
    int SortOrder,
    byte[]? RowVersion);

public sealed record CancelPrescriptionRequest(string? Reason);

public sealed record SearchPrescriptionsRequest(
    Guid? PatientId,
    Guid? DoctorId,
    Guid? MedicalVisitId,
    string? PrescriptionNumber,
    DateOnly? DateFrom,
    DateOnly? DateTo,
    string? Status,
    int Page = 1,
    int PageSize = 20);

public sealed record PatientPrescriptionHistoryRequest(
    DateOnly? DateFrom,
    DateOnly? DateTo,
    string? Status,
    int Page = 1,
    int PageSize = 20);

public sealed record PrescriptionItemDto(
    Guid Id,
    Guid MedicationId,
    string MedicationNameSnapshot,
    string Dosage,
    string Frequency,
    string? Duration,
    string? Route,
    string? Instructions,
    decimal? Quantity,
    string? Notes,
    int SortOrder);

public sealed record PrescriptionDto(
    Guid Id,
    Guid OrganizationId,
    Guid BranchId,
    string PrescriptionNumber,
    Guid MedicalVisitId,
    Guid PatientId,
    Guid DoctorId,
    DateOnly PrescriptionDate,
    string Status,
    string? Notes,
    DateTime? IssuedAtUtc,
    string? IssuedBy,
    DateTime? CancelledAtUtc,
    string? CancelledBy,
    string? CancellationReason,
    DateTime CreatedAtUtc,
    string? CreatedBy,
    DateTime? UpdatedAtUtc,
    string? UpdatedBy,
    byte[] RowVersion,
    IReadOnlyList<PrescriptionItemDto> Items);

public sealed record PrescriptionListItemDto(
    Guid Id,
    string PrescriptionNumber,
    DateOnly PrescriptionDate,
    Guid PatientId,
    Guid DoctorId,
    Guid MedicalVisitId,
    string Status,
    int ItemCount);

public sealed record PagedPrescriptionsResult(
    IReadOnlyList<PrescriptionListItemDto> Items,
    int Page,
    int PageSize,
    int TotalCount);

File: frontend/erpclink-web/src/app/features/prescriptions/models/prescription.models.ts
export type PrescriptionStatus = 'Draft' | 'Issued' | 'Cancelled';

export interface PrescriptionItemDto {
  id: string;
  medicationId: string;
  medicationNameSnapshot: string;
  dosage: string;
  frequency: string;
  duration: string | null;
  route: string | null;
  instructions: string | null;
  quantity: number | null;
  notes: string | null;
  sortOrder: number;
}

export interface PrescriptionDto {
  id: string;
  organizationId: string;
  branchId: string;
  prescriptionNumber: string;
  medicalVisitId: string;
  patientId: string;
  doctorId: string;
  prescriptionDate: string;
  status: PrescriptionStatus;
  notes: string | null;
  issuedAtUtc: string | null;
  issuedBy: string | null;
  cancelledAtUtc: string | null;
  cancelledBy: string | null;
  cancellationReason: string | null;
  createdAtUtc: string;
  createdBy: string | null;
  updatedAtUtc: string | null;
  updatedBy: string | null;
  rowVersion: string;
  items: PrescriptionItemDto[];
}

export interface PrescriptionListItemDto {
  id: string;
  prescriptionNumber: string;
  prescriptionDate: string;
  patientId: string;
  doctorId: string;
  medicalVisitId: string;
  status: PrescriptionStatus;
  itemCount: number;
}

export interface PagedPrescriptionsResult {
  items: PrescriptionListItemDto[];
  page: number;
  pageSize: number;
  totalCount: number;
}

export interface PrescriptionItemInput {
  medicationId: string;
  dosage: string;
  frequency: string;
  duration?: string | null;
  route?: string | null;
  instructions?: string | null;
  quantity?: number | null;
  notes?: string | null;
  sortOrder?: number | null;
}

export interface CreatePrescriptionRequest {
  medicalVisitId: string;
  notes?: string | null;
  items?: PrescriptionItemInput[] | null;
}

export interface UpdatePrescriptionRequest {
  notes?: string | null;
  rowVersion?: string | null;
}

export interface UpdatePrescriptionItemRequest {
  dosage: string;
  frequency: string;
  duration?: string | null;
  route?: string | null;
  instructions?: string | null;
  quantity?: number | null;
  notes?: string | null;
  sortOrder: number;
  rowVersion?: string | null;
}

export interface CancelPrescriptionRequest {
  reason?: string | null;
}

export interface SearchPrescriptionsParams {
  patientId?: string;
  doctorId?: string;
  medicalVisitId?: string;
  prescriptionNumber?: string;
  dateFrom?: string;
  dateTo?: string;
  status?: PrescriptionStatus | '';
  page?: number;
  pageSize?: number;
}

export interface PatientPrescriptionHistoryParams {
  dateFrom?: string;
  dateTo?: string;
  status?: PrescriptionStatus | '';
  page?: number;
  pageSize?: number;
}

export function isPrescriptionDraft(status: PrescriptionStatus): boolean {
  return status === 'Draft';
}

export function isPrescriptionEditable(status: PrescriptionStatus | string): boolean {
  return status === 'Draft';
}

export function canIssuePrescription(status: PrescriptionStatus | string, itemCount: number): boolean {
  return status === 'Draft' && itemCount > 0;
}

export function canCancelPrescription(status: PrescriptionStatus | string): boolean {
  return status === 'Draft' || status === 'Issued';
}
